### ISSUE TITLE
gamemodel never starts (GameModel.run() isn't entered) sometimes

### ISSUE THREAD

**Issue by pychess** [pychess/pychess#883]
Original [issue 883](https://code.google.com/p/pychess/issues/detail?id=883) reported by [mattgatto](https://code.google.com/u/mattgatto/) 2014-06-11

I was doing some testing on FICS by following a player who was playing lots of lightning matches and after about 4 different game tabs with respective analyzer instances (stockfish) had been opened (creating alot of CPU overhead and apparently altering timings enough to manifest this bug), and upon the player starting a new match, the gamewidget was created and attached and the board was drawn, but the pieces weren't drawn and the game didn't start.

I think this is caused by a race condition present in ThreadPool.py. My reasoning is based on the attached log, in which ionest.workfunc finishes according to the log, and therefore, GameModel.loadAndStart must have completed and GameModel.start() must have been called. Also, the next thread dump after the game is unsuccessfully started shows this thread which doesn't show up previously:

Thread: ICGameModel.run:537 (140291273123584)
  File "/usr/lib64/python2.7/threading.py", line 784, in __bootstrap
    self.__bootstrap_inner()
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
    self.run()
  File "/home/gatto/code/hg/pychess/lib/pychess/System/ThreadPool.py", line 103, in run
    self.wcond.wait()    # wait for work
  File "/usr/lib64/python2.7/threading.py", line 339, in wait
    waiter.acquire()

So it appears that GameModel.run is called and the thread is initialized (via 'a.func = ...') and named in ThreadPool.start, yet the following first line of GameModel.run is never executed:
        log.debug("GameModel.run: Starting. self=%s" % self)

The only explanation I can think of is a race condition where, after the ThreadPool.Worker finishes running its self.func and puts itself back into the thread pool via:

```
                    self.worker = None
                    self.queue.put(self)
```

... the code in ThreadPool.start below:

```
    a.worker = lambda: worker(*args, **kw)
    a.name = worker.name

    a.wcond.acquire()
    a.wcond.notify()
    a.wcond.release()
```

... runs before the code in ThreadPool.Worker.run below does:

```
                self.wcond.acquire()
                self.wcond.wait()    # wait for work
                self.wcond.release()
```

... Which would cause the Worker to sit endlessly waiting for a wcond.notify that never will occur.

**Comment by pychess** [comment]
- attached [2014-06-07_14-31-55-truncated.log.bz2](https://code.google.com/p/pychess/issues/detail?id=883#c0)

**Comment by pychess** [comment]
Original [comment 1.](https://code.google.com/p/pychess/issues/detail?id=883#c1) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-06-11
- added label: Critical

**Comment by pychess** [comment]
Original [comment 2.](https://code.google.com/p/pychess/issues/detail?id=883#c2) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-06-12

I attached a patch that resolves this by getting rid of the ThreadPool altogether. Does anybody see a reason to keep it? We don't have anything that requires a thread pool, like a server or heavy computations.
- attached [patch-1000.diff](https://code.google.com/p/pychess/issues/detail?id=883#c2)
- added label: Anderssen0.12

**Comment by pychess** [comment]
Original [comment 3.](https://code.google.com/p/pychess/issues/detail?id=883#c3) posted by [gbtami](https://code.google.com/u/gbtami/) 2014-06-13

I know no reason to keep ThreadPool at all.

Just one little note:
in GtkWorker.py line comment 83.
        Thread.__init__(self, fident(func))
should be
        Thread.__init__(self, name=fident(func))

**Comment by pychess** [comment]
Original [comment 4.](https://code.google.com/p/pychess/issues/detail?id=883#c4) posted by [mattgatto](https://code.google.com/u/mattgatto/) 2014-06-20

This issue was closed by revision 3c86505c38fc.
- changed status to: Fixed
